### PR TITLE
Add optional generative model updates to variational inference functions

### PR DESCRIPTION
**Pull request by marcoct**
This PR adds the optional ability to update the parameters of the generative model within the `black_box_vi!` and `black_box_vimco!` functions, so these can be used to implement updates to both the variational family and generative model. Previously, the user would have needed to implement the generative model parameter updates separately, interleaved with one iteration of the variational family update.

### CODE REVIEW

**Comment by marcoct** [comment]
I'm marking this as a draft, since I have not added unit tests for this new functionality yet.

(Thanks to @ztangent  for bringing this issue to attention and @alex-lew for suggesting this change)

**Review by ztangent — COMMENTED** [review]
This is looking great! I just have two main questions in my review:

1. Whether we should redesign the method signature at avoid long keyword argument names.
2. Whether the VIMCO model update should make use of multiple model traces, instead of just one.

**Comment by ztangent** [comment]
Also, it looks like this PR needs to be updated to fix some conflicts!

File: src/inference/variational.jl [hunk 1]
```diff
@@ -88,95 +88,132 @@ function multi_sample_gradient_estimate!(
...
 """
     (elbo_estimate, traces, elbo_history) = black_box_vi!(
         model::GenerativeFunction, args::Tuple,
         observations::ChoiceMap,
         var_model::GenerativeFunction, var_model_args::Tuple,
-        update::ParamUpdate;
-        iters=1000, samples_per_iter=100, verbose=false)
+        variational_family_update::ParamUpdate;
+        iters=1000, samples_per_iter=100, verbose=false,
+        generative_model_update::Union{ParamUpdate,Nothing}=nothing)
```
[the remark this answers is not in the document]
Comment (ztangent) [reply src/inference/variational.jl]: Made these changes in my recent commits.

File: src/inference/variational.jl [hunk 2]
```diff
@@ -140,7 +144,7 @@ function black_box_vi!(
             elbo_estimate += (log_weight / samples_per_iter)
 
             # accumulate the generative model gradients
-            _maybe_accumulate_param_grad!(model_trace, model_update)
+            _maybe_accumulate_param_grad!(model_trace, model_update, 1.0 / samples_per_iter)
```
Comment (marcoct) [inline src/inference/variational.jl:147]: Previously, changing the number of samples changed the magnitude of the gradient and would have required a reciprocal change to the step size to maintain similar behavior. This fixes this.

**Comment by ztangent** [comment]
@marcoct it looks like this PR is now failing some tests? :(

**Pull request merged**